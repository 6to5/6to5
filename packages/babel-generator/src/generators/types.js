/* eslint max-len: 0 */
/* eslint quotes: 0 */

import * as t from "babel-types";

export function Identifier(node: Object) {
  // FIXME: We hang variance off Identifer to support Flow's def-site variance.
  // This is a terrible hack, but changing type annotations to use a new,
  // dedicated node would be a breaking change. This should be cleaned up in
  // the next major.
  if (node.variance) {
    if (node.variance === "plus") {
      this.token("+");
    } else if (node.variance === "minus") {
      this.token("-");
    }
  }

  this.word(node.name);
}

export function RestElement(node: Object) {
  this.token("...");
  this.print(node.argument, node);
}

export {
  RestElement as SpreadElement,
  RestElement as SpreadProperty,
  RestElement as RestProperty,
};

export function ObjectExpression(node: Object) {
  let props = node.properties;

  this.token("{");
  this.printInnerComments(node);

  if (props.length) {
    this.space();
    this.printList(props, node, { indent: true, statement: true });
    this.space();
  }

  this.token("}");
}

export { ObjectExpression as ObjectPattern };

export function ObjectMethod(node: Object) {
  this.printJoin(node.decorators, node);
  this._method(node);
}

export function ObjectProperty(node: Object) {
  this.printJoin(node.decorators, node);

  if (node.computed) {
    this.token("[");
    this.print(node.key, node);
    this.token("]");
  } else {
    // print `({ foo: foo = 5 } = {})` as `({ foo = 5 } = {});`
    if (t.isAssignmentPattern(node.value) && t.isIdentifier(node.key) && node.key.name === node.value.left.name) {
      this.print(node.value, node);
      return;
    }

    this.print(node.key, node);

    // shorthand!
    if (node.shorthand &&
      (t.isIdentifier(node.key) &&
       t.isIdentifier(node.value) &&
       node.key.name === node.value.name)) {
      return;
    }
  }

  this.token(":");
  this.space();
  this.print(node.value, node);
}

export function ArrayExpression(node: Object) {
  let elems = node.elements;
  let len   = elems.length;

  this.token("[");
  this.printInnerComments(node);

  for (let i = 0; i < elems.length; i++) {
    let elem = elems[i];
    if (elem) {
      if (i > 0) this.space();
      this.print(elem, node);
      if (i < len - 1) this.token(",");
    } else {
      // If the array expression ends with a hole, that hole
      // will be ignored by the interpreter, but if it ends with
      // two (or more) holes, we need to write out two (or more)
      // commas so that the resulting code is interpreted with
      // both (all) of the holes.
      this.token(",");
    }
  }

  this.token("]");
}

export { ArrayExpression as ArrayPattern };

export function RegExpLiteral(node: Object) {
  this.word(`/${node.pattern}/${node.flags}`);
}

export function BooleanLiteral(node: Object) {
  this.word(node.value ? "true" : "false");
}

export function NullLiteral() {
  this.word("null");
}

export function NumericLiteral(node: Object) {
  let raw = this.getPossibleRaw(node);

  this.number(raw == null ? node.value + "" : raw);
}

export function StringLiteral(node: Object, parent: Object) {
  // raw is the string literal's representation as it occurs
  // in the source code. This preserves quotes, line break
  // characters and unicode code units.
  let val = this.getPossibleRaw(node);
  // fall back to stringified value if raw does not exist
  if (val == null) {
    val = JSON.stringify(node.value);
  }

  // escape illegal js but valid json unicode characters
  val = val.replace(/[\u000A\u000D\u2028\u2029]/g, function (c) {
    return "\\u" + ("0000" + c.charCodeAt(0).toString(16)).slice(-4);
  });

  // Early return when inside JSX, because quotes
  // should not be altered in this case.
  if (t.isJSX(parent)) {
    return this.token(val);
  }

  // remove current quotes (keep originalVal for comparison below)
  let originalVal = val;
  val = val.slice(1, -1);

  if (this.format.quotes === "single") {
    // unescape double quotes
    val = val.replace(/\\"/g, '"');

    // escape single quotes not already escaped
    val = val.replace(/\\([\s\S])|(')/g,"\\$1$2");

    // add single quotes
    val = `'${val}'`;
  }

  // double quotes - default
  if (this.format.quotes !== "single") {
    // unescape single quotes
    val = val.replace(/\\'/g, "'");

    // escape double quotes not already escaped
    val = val.replace(/\\([\s\S])|(")/g,"\\$1$2");

    // add double quotes
    val = `"${val}"`;
  }

  // If conversion to led to escaping quotes and
  // original string had no escaped quotes, prefer
  // the original if avoidEscapingQuotes is true.
  if (this.format.avoidEscapingQuotes
    && /\\'|\\"/.test(val)
    && !/\\'|\\"/.test(originalVal)) {
    return this.token(originalVal);
  }

  return this.token(val);
}
