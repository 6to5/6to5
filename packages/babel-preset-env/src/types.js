//@flow

import { TargetNames, ModulesOption, InjectPolyfillsOption } from "./options";

// Targets
export type Target = $Keys<typeof TargetNames>;
export type Targets = {
  [target: Target]: string,
};

// Options
// Use explicit modules to prevent typo errors.
export type ModuleOption = $Values<typeof ModulesOption>;
export type PolyfillsOption = $Values<typeof InjectPolyfillsOption>;

export type Options = {
  configPath: string,
  debug: boolean,
  exclude: Array<string | RegExp>,
  forceAllTransforms: boolean,
  ignoreBrowserslistConfig: boolean,
  include: Array<string | RegExp>,
  loose: boolean,
  modules: ModuleOption,
  shippedProposals: boolean,
  spec: boolean,
  targets: Targets,
  useBuiltIns: boolean,
  injectPolyfills: PolyfillsOption,
};

// Babel
export type Plugin = [Object, Object];
