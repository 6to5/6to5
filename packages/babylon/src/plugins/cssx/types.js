import { TokenType, types as tt } from "../../tokenizer/types";
import { TokContext, types as tc } from "../../tokenizer/context";

tt.cssxSelector = new TokenType('cssxSelector');
tt.cssxRulesStart = new TokenType('cssxRulesStart');
tt.cssxRulesEnd = new TokenType('cssxRulesEnd');
tt.cssxProperty = new TokenType('cssxProperty');
tt.cssxValue = new TokenType('cssxValue');
tt.cssxMediaQuery = new TokenType('cssxMediaQuery');
tt.cssxMediaQueryStart = new TokenType('cssxMediaQueryStart');
tt.cssxMediaQueryEnd = new TokenType('cssxMediaQueryEnd');

tt.cssxRulesStart.updateContext = function (prevType) {
  if (prevType === tt.cssxSelector) this.state.context.push(tc.cssxRules);
};
tt.cssxRulesEnd.updateContext = function (prevType) {
  if (
    prevType === tt.cssxValue ||
    prevType === tt.cssxRulesStart ||
    prevType === tt.semi
  ) {
    this.state.context.length -= 1; // out of cssxRules
    this.cssxOut();
  }
};
tt.cssxSelector.updateContext = function (prevType) {
  this.state.context.length -= 1;
};