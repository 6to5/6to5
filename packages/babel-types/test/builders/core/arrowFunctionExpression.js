import * as t from "@babel/types";

describe("builders", function () {
  describe("es2015", function () {
    describe("arrowFunctionExpression", function () {
      it("accept all parameters", function () {
        const arrowFunctionExpression = t.arrowFunctionExpression(
          [t.identifier("foo")],
          t.blockStatement([t.returnStatement(t.thisExpression())]),
          true,
          true,
          true,
          t.tsTypeAnnotation(t.tsStringKeyword()),
          t.tsTypeParameterDeclaration([
            t.tsTypeParameter(t.tsObjectKeyword(), null, "foo"),
          ]),
        );
        expect(arrowFunctionExpression).toMatchSnapshot();
      });
    });
  });
});
