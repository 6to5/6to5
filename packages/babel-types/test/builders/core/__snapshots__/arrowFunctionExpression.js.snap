// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builders es2015 arrowFunctionExpression accept all parameters 1`] = `
Object {
  "async": true,
  "body": Object {
    "body": Array [
      Object {
        "argument": Object {
          "type": "ThisExpression",
        },
        "type": "ReturnStatement",
      },
    ],
    "directives": Array [],
    "type": "BlockStatement",
  },
  "expression": true,
  "generator": true,
  "params": Array [
    Object {
      "name": "foo",
      "type": "Identifier",
    },
  ],
  "returnType": Object {
    "type": "TSTypeAnnotation",
    "typeAnnotation": Object {
      "type": "TSStringKeyword",
    },
  },
  "type": "ArrowFunctionExpression",
  "typeParameters": Object {
    "params": Array [
      Object {
        "constraint": Object {
          "type": "TSObjectKeyword",
        },
        "default": null,
        "name": "foo",
        "type": "TSTypeParameter",
      },
    ],
    "type": "TSTypeParameterDeclaration",
  },
}
`;
